Enable nullable annotations for M.E.L.Abstractions

maryamariyan opened the pull request:
cc: @buyaa-n @safern 

Related to https://github.com/dotnet/runtime/issues/43605

maryamariyan committed:
Enable nullable annotations for M.E.L.Abstractions

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExternalScopeProvider.cs:
```diff
@@ -49,18 +49,18 @@ private class Scope : IDisposable
-            public Scope Parent { get; }
+            public Scope? Parent { get; }
 
             public object State { get; }
```
- [ ] Perhaps State is nullable too.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogEntry.cs:
```diff
@@ -19,7 +19,7 @@ namespace Microsoft.Extensions.Logging.Abstractions
         /// <param name="exception">The log exception.</param>
         /// <param name="formatter">The formatter.</param>
-        public LogEntry(LogLevel logLevel, string category, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        public LogEntry(LogLevel logLevel, string category, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
```
- [ ] Perhaps formatter can be nullable too

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/src/FormattedLogValues.cs:
```diff
@@ -66,10 +66,10 @@ public FormattedLogValues(string format, params object[] values)
                 }
 
-                return _formatter.GetValue(_values, index);
+                return _formatter!.GetValue(_values!, index);
```
Note for reviewer: _values is not null due to what `Count` returns on line 76

maryamariyan commented on src/libraries/Common/src/Extensions/TypeNameHelper/TypeNameHelper.cs:
```diff
@@ -110,14 +110,14 @@ private static void ProcessGenericType(StringBuilder builder, Type type, Type[]
             if (type.IsNested)
             {
-                offset = type.DeclaringType.GetGenericArguments().Length;
+                offset = type.DeclaringType!.GetGenericArguments().Length;
```
https://docs.microsoft.com/en-us/dotnet/api/system.type.declaringtype?view=net-5.0

`type.DeclaringType` is not null when Nested or IsArray

maryamariyan committed:
Corrections from review

krwq commented on src/libraries/Common/src/Extensions/Logging/NullExternalScopeProvider.cs:
```diff
@@ -20,12 +22,12 @@ private NullExternalScopeProvider()
 
         /// <inheritdoc />
-        void IExternalScopeProvider.ForEachScope<TState>(Action<object, TState> callback, TState state)
+        void IExternalScopeProvider.ForEachScope<TState>(Action<object?, TState> callback, TState state)
```
should TState be nullable?

stephentoub replied:
> should TState be nullable?

It looks fine as is.  If a developer wants to provide a nullable TState, they can.  I don't see a benefit here in constraining it to be notnull.

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs:
```diff
@@ -50,41 +50,41 @@ public partial interface ISupportExternalScope
-        public static void LogWarning(this Microsoft.Extensions.Logging.ILogger logger, Microsoft.Extensions.Logging.EventId eventId, string message, params object[] args) { }
-        public static void LogWarning(this Microsoft.Extensions.Logging.ILogger logger, System.Exception exception, string message, params object[] args) { }
-        public static void LogWarning(this Microsoft.Extensions.Logging.ILogger logger, string message, params object[] args) { }
+        public static System.IDisposable BeginScope(this Microsoft.Extensions.Logging.ILogger logger, string messageFormat, params object?[]? args) { throw null; }
```
It's generally strange to see `params` arrays be nullable, i.e. I'd have expected this to be `params object?[]` rather than `params object?[]?`.  Are there known uses that pass in a null array?  In fact, it looks like the implementation explicitly checks for args being null and throws if it is.

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs:
```diff
@@ -50,41 +50,41 @@ public partial interface ISupportExternalScope
-        public static void LogWarning(this Microsoft.Extensions.Logging.ILogger logger, System.Exception exception, string message, params object[] args) { }
-        public static void LogWarning(this Microsoft.Extensions.Logging.ILogger logger, string message, params object[] args) { }
+        public static System.IDisposable BeginScope(this Microsoft.Extensions.Logging.ILogger logger, string messageFormat, params object?[]? args) { throw null; }
+        public static void Log(this Microsoft.Extensions.Logging.ILogger logger, Microsoft.Extensions.Logging.LogLevel logLevel, Microsoft.Extensions.Logging.EventId eventId, System.Exception? exception, string? message, params object?[]? args) { }
```
From looking at the implementation, it does technically look like args can be null, but do we see anyone actually doing that?  I'd be tempted to say that all of these params arrays should be non-nullable (though allowing the objects they contain to be nullable), as that's generally what we've done elsewhere, and explicitly passing in a null array would be discouraged.

maryamariyan replied:
Agreed. So based on this sample:

https://github.com/dotnet/runtime/blob/4d09ba6b7e458b491d0ba3d8b8d2a0ac0a289a09/src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerTest.cs#L725-L728

Which would be calling into:

https://github.com/dotnet/runtime/blob/4d09ba6b7e458b491d0ba3d8b8d2a0ac0a289a09/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExtensions.cs#L410-L421

Would make sense to make args non-nullable but the the objects themselves be allowed null.

I'll push up a commit shortly to switch all `params object?[]? args` instances in LoggerExtensions.cs to `params object?[] args`

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs:
```diff
@@ -133,11 +133,11 @@ namespace Microsoft.Extensions.Logging.Abstractions
         private readonly int _dummyPrimitive;
-        public LogEntry(Microsoft.Extensions.Logging.LogLevel logLevel, string category, Microsoft.Extensions.Logging.EventId eventId, TState state, System.Exception exception, System.Func<TState, System.Exception, string> formatter) { throw null; }
+        public LogEntry(Microsoft.Extensions.Logging.LogLevel logLevel, string category, Microsoft.Extensions.Logging.EventId eventId, TState state, System.Exception? exception, System.Func<TState, System.Exception?, string>? formatter) { throw null; }
         public string Category { get { throw null; } }
```
Structs like this are tough.  While the constructor declaring `string category` as non-nullable means we wouldn't expect any nulls here, it is possible for someone to do `LogEntry<TState> le = default;` or `LogEntry<TState> le = new LogEntry<TState>()` (which are equivalent), in which case `le.Category` will be null.  So, this comes down to a judgement call as to how common it'll be for someone to work with LogEntry's, might they use them zero-initialized, etc.  If the answer is "No, the 99.9% case is a developer only uses instances produced by the framework, and those will always have been constructed correctly before being handed out", then the current annotation looks good.  If instead the answer is "Yes, there's a non-negligible chance a developer will find themselves working with a LogEntry that may have its default value", then Category and State should both be marked `?` as well.

maryamariyan replied:
In the repo today, LogEntry is only used for console logging and specifically on the `ConsoleFormatter.Write<TState>` API which receives a LogEntry instance. As you also mentioned, it is expected that mostly the developer just uses the instance provided/instantiated by the framework on the first line below:

https://github.com/dotnet/runtime/blob/23611e16a93c9056ceeb27c55e93809e1baf56c4/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L45-L46

and that gets passed onto a console formatter in the next line, which may or may not be provided by the framework.

In a situation where (in the future) this pattern gets adopted by other loggers (e.g. DebugLogger, and not just ConsoleLogger), then what would happen is that `ILogger.Log<TState>(logLevel, eventId, state, exception, formatter)` would be responsible for instantiating and passing down LogEntry to a `Formatter.Write<TState>(logEntry, ...)` API which ends up just using it and doing custom formatting.

To better understand whether or not LogEntry properties should be nullable and I looked at more usages of:
```c#
ILogger.Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter);
```
in the framework. 

- [x] **`TState state`** and **`Exception? exception`**: up to caller to pass null. OK as nullable.
- [x] **`LogLevel logLevel`** is an enum
- [x] **`EventId eventId`**  is a struct
- [x] **`string category`**: I believe category should not be nullable. 

e.g for ConsoleLogger
https://github.com/dotnet/runtime/blob/5e5c4ebf1df276103b5a923a14325f845c2e7177/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L45
`_name` cannot be null.

Also checked LoggerFactory in Microsoft.Extensions.Logging which helps setup different LoggerProviders (ConsoleLogger being one of them), looking at its `CreateLogger(string categoryName)` API below, confirmed it would also throw ArgumentNullException when null category passed in.
so something like 
```c#
public static async Task Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder => { builder.AddConsole();  });

            var logger = loggerFactory.CreateLogger(categoryName); // when null throws ArgumentNullException
            logger.Log(LogLevel.Information, 0, new Exception(), null, (t, s) => string.Empty);
        }
```

- [ ] **`Func<TState, Exception?, string> formatter`**: Did a similar investigation for the formatter argument, and I believe formatter should not be nullable either,
For EventLogLogger, ConsoleLogger and EventSourceLogger, passing null formatter throws exception. 

There is also a NullLogger that is a noop on Log, therefore not throwing ArgumentNullException. And TraceSourceLogger allows null formatter as seen below (handles it rather than throwing):

https://github.com/dotnet/runtime/blob/5cb5ae9f33837ec63e1a444f5d8a3f64d2d7188b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs#L28-L35

Would it be reasonable to restrict `formatter` argument in `ILogger.Write<TState>(.., formatter)` to be not nullable and make TraceSourceLogger consistent with the others through this change? 

---

Back to your question on "how common it would be for someone to work with LogEntry's", I think it was just introduced on 5.0 to conveniently pass down ILogger.Log<TState> arguments onto Formatter.Write<TState> and for that it is not expected that someone would get use of default `LogEntry<TState> le = default;` or `LogEntry<TState> le = new LogEntry<TState>()`.

maryamariyan replied:
Chatted offline, the conclusion is to have ILogger.Log's formatter argument to be non-nullable

krwq commented on src/libraries/Common/src/Extensions/TypeNameHelper/TypeNameHelper.cs:
```diff
@@ -31,7 +33,7 @@ internal static class TypeNameHelper
         };
 
-        public static string GetTypeDisplayName(object item, bool fullName = true)
+        public static string? GetTypeDisplayName(object? item, bool fullName = true)
```
should this be `[return: NotNullIfNotNull("item")]`?

krwq replied:
(also if it has impact on other places then follow it)

krwq commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs:
```diff
@@ -10,11 +10,11 @@ namespace Microsoft.Extensions.Logging
     {
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
-        public EventId(int id, string name = null) { throw null; }
```
please make sure owner of this library double checks the annotations of the ref

krwq commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs:
```diff
@@ -23,14 +23,14 @@ namespace Microsoft.Extensions.Logging
     {
-        void ForEachScope<TState>(System.Action<object, TState> callback, TState state);
-        System.IDisposable Push(object state);
+        void ForEachScope<TState>(System.Action<object?, TState> callback, TState state);
```
should `TState` be nullable?

maryamariyan replied:
Same as https://github.com/dotnet/runtime/pull/43892#discussion_r514370522

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/src/Microsoft.Extensions.Logging.Abstractions.csproj:
```diff
@@ -1,8 +1,12 @@
+    <ExcludeCurrentNetCoreAppFromPackage>true</ExcludeCurrentNetCoreAppFromPackage>
     <EnableDefaultItems>true</EnableDefaultItems>
+    <!-- Use targeting pack references instead of granular ones in the project file. -->
+    <DisableImplicitAssemblyReferences>false</DisableImplicitAssemblyReferences>
```
Did you intend to check in these changes?

maryamariyan replied:
yes this was intentional otherwise wouldn't have been able to add NetCoreAppCurrent for the project.

cc: @safern

safern replied:
The reason behind adding `NetCoreAppCurrent` for these projects was to have CI protection for the nullable annotations to actually be correct and not broken on by upcoming changes but to actually not ship that asset. However the only caveat that this brings is that by adding this I'm not sure if people can just clone the repo and build this project without pre building anything else which was an effort @ViktorHofer did in the past.

ViktorHofer replied:
> yes this was intentional otherwise wouldn't have been able to add NetCoreAppCurrent for the project I think.

This property is for convenience only, in case you don't want to list all the individual assembly references. In this case I think it's fine as we don't ship the `$(NetCoreAppCurrent)' asset.

> However the only caveat that this brings is that by adding this I'm not sure if people can just clone the repo and build this project without pre building anything

Unfortunately with our current infra that's true but I don't think we should see that as a blocker. I would like to improve our infra so that `$(NetCoreAppCurrent)' is able to build via dependencies without an up-front build which would also enable VS builds for .NETCoreApp without up-front steps.

stephentoub reviewed (commented):
Overall LGTM.

krwq commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogEntry.cs:
```diff
@@ -19,7 +19,7 @@ namespace Microsoft.Extensions.Logging.Abstractions
         /// <param name="exception">The log exception.</param>
         /// <param name="formatter">The formatter.</param>
-        public LogEntry(LogLevel logLevel, string category, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        public LogEntry(LogLevel logLevel, string category, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string>? formatter)
```
TState seems it may be null. Double check with the documentation

stephentoub replied:
It's up to the caller whether they want to specify a nullable or non-nullable TState.

krwq reviewed (approved):
LGTM with some questions

maryamariyan committed:
LoggerExtensions.cs: params object?[] -> params object?[]

maryamariyan committed:
using Nullable Annotations,
to allow annotations on relevant projects, not yet enabled

maryamariyan committed:
formatter not null

maryamariyan commented:
@stephentoub @krwq @tarekgh looked at all the comments and made all the changes I was planning for.

I posed this question in https://github.com/dotnet/runtime/pull/43892#discussion_r518056956 too. 
> Would it be reasonable to restrict formatter argument in ILogger.Write<TState>(.., formatter) to be not nullable and make TraceSourceLogger consistent with the others through this change? (refer to API diff for making formatter not nullable)

let me know what you think.

maryamariyan commented:
Updates on PR since it was last reviewed:

- [x] Addressed PR feedback
- [x] LoggerExtensions.cs: params object?[] -> params object?[]
- [x] Last commit makes formatter non-nullable

maryamariyan commented:
@stephentoub @krwq could you please re-review?

maryamariyan commented:
perf legs usually don't run on PRs normally wonder why they showed up here.

cc: @ericstj

ViktorHofer commented:
cc @DrewScoggins

krwq commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/Microsoft.Extensions.DependencyInjection.csproj:
```diff
@@ -4,6 +4,7 @@
     <TargetFrameworks>$(NetCoreAppCurrent);net461;netstandard2.0;netstandard2.1</TargetFrameworks>
     <!-- Debug IL generation -->
     <ILEmitBackendSaveAssemblies>False</ILEmitBackendSaveAssemblies>
+    <Nullable>Annotations</Nullable>
```
are you planning to also enable warnings?

maryamariyan replied:
The reason Annotation is enabled is because DI and Http use common shared files also used by Logging.Abstractions (NullExternalScopeProvider.cs and TypeNameHelper.cs) and they both have
```
#nullable enable
```
to them.

maryamariyan committed:
Merge remote-tracking branch 'upstream/master' into logging-abstractions-null

maryamariyan committed:
Merge remote-tracking branch 'upstream/master' into logging-abstractions-null

Conflicts:
	src/libraries/Microsoft.Extensions.Logging.Abstractions/src/Microsoft.Extensions.Logging.Abstractions.csproj

stephentoub commented:
@maryamariyan, doesn't the ref csproj need `<Nullable>enable</Nullable>` as well?
